### PR TITLE
Allow brave_domain callers to specify default env for unofficial builds

### PR DESCRIPTION
**Pull request by nvonpentz**
Resolves https://github.com/brave/brave-browser/issues/35497

Update GetServicesDomain
* Makes prefix a required parameter (wasn't being used, and made the typing easier)
* Add new optional parameter env_value_default_override that is used as an override to the default environment in non official builds

This allows us to optionally specify a different default environment for each GetServicesDomain call in unofficial builds. E.g. For AI chat, we specify dev and staging for aichat and leo domains respectively.  Of course, these can be overriden by the flags, if passed.

// Precedence is:
// 1. Prefix specific CLI overrides
// 2. Global CLI overrides overrides
// 3. Default env override parameter
// 4. Default env (production)

Resolves https://github.com/brave/brave-browser/issues/35497

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: brave_domains/service_domains.h
Comment (petemill) [inline brave_domains/service_domains.h:37]: I wonder if this should be an enum. As it's a string, it's not obvious to the caller what the value they should put in. They might type it out themselves. We could put an enum here (e.g. `ServicesEnvironment::Dev|Staging|Production`) or at least have a `DCHECK` if it's not one of the expected values. ...but that doesn't stop people putting the string in directly...

**Review by petemill — APPROVED** [review]
Approving to unblock. I think it's fine the way it is, but do think the enum could be an improvement. If that works out in this PR then great I'll review again quickly before merge, otherwise you could do as an immediate follow-up.

File: brave_domains/service_domains.h
Comment (petemill) [reply brave_domains/service_domains.h:37]: Also I wonder if this should default to dev for everything? I guess I don't understand what the purpose of dev vs staging is - there's possibilities dev instances have in-progress code that isn't suitable for master? But if that's the case then ai chat will suffer...?

File: brave_domains/service_domains.h
Comment (petemill) [inline brave_domains/service_domains.h:15]: According to the chromium c++ style guide, these should be in the format `DEV`, `STAGING` and `PROD` (should it be `PRODUCTION`?)

File: brave_domains/service_domains_unittest.cc
Comment (petemill) [inline brave_domains/service_domains_unittest.cc:45]: nit: would rather see the enum type too, i.e. `ServicesEnvironment::PROD` here and the other uses

File: brave_domains/service_domains.cc
Comment (petemill) [inline brave_domains/service_domains.cc:76]: nit: DCHECK or NOTREACHED here, because we shouldn't be in this point as we have every value of the enum covered

**Review by petemill — APPROVED** [review]
Still looks good, aside from the one nit

File: brave_domains/service_domains.cc
Comment (bridiver) [inline brave_domains/service_domains.cc:87]: specifying a default when you call `GetServicesDomain` seems extremely fragile/error prone to me and could result in different parts of a feature having different default envs in non official builds. My understanding is that the problem is that some services only have two environments in which case we should just map dev and staging to the same thing

Comment (bridiver) [reply brave_domains/service_domains.cc:87]: we can do it at the dns level

**Pull request merged**
